### DISCUSSION TITLE
v0.4.0

### DISCUSSION THREAD

**Discussion by hjiangsu** [thunder-app/thunder#1399] [Announcements]
Hey everyone, Thunder v0.4.0 is finally here! This release cycle took a bit longer than usual, but as a result, introduces a lot of new features, changes, and fixes which should make for a even better Thunder experience. If you have any suggestions or feedback, feel free to discuss it in the Thunder community or on GitHub.

If you're enjoying Thunder and would like to show some support, star this project on GitHub! This helps gain visibility for the project and allows more contributors to help with the continued development of Thunder. If you'd like to show some appreciation, [here is a post with details on how to donate](https://github.com/thunder-app/thunder/issues/756) to the active contributors of Thunder.

A quick reminder: if you are using the Google Play version or App Store version, please note that it may take a bit of time before you receive the update. Now onto the update notes. This update brings a few major features:

🎉 New Features
- The comment creation page has been fully re-built from the ground up and introduces a few new features, including the ability to specify the comment language, ability to comment as a different user, and addition of spoilers to the markdown toolbar. If you have any suggestions or feedback regarding these changes, feel free to make a new issue on GitHub!
- There is now initial support for video playback on Thunder! The built-in video player should work with most video URLs (that end in a proper video extension) and should work with YouTube URLs. Additionally, you can customize some video settings such as playback speed, auto-mute, auto-play and more in the Settings page. The video player is still experimental, so there may still be some issues present.
- Additional customization options including the ability to customize the colours of actions (upvotes, downvotes, saving, etc.), hiding thumbnails from the feed, showing user/community avatars, and much more.

ℹ General
- There have been improvements to the overall UI/UX across the board. This includes improvements to the pure black theme, improvements to markdown rendering of tables, quotes, and more.
- There have also been a lot of changes under-the-hood to hopefully improve the overall performance and reliability of actions. As a result of these changes, there may still be some issues that were not caught during testing. If you notice any general issues, please make a new issue on GitHub.

And much much more. I can't go through all the changes in the changelog, but for those who are curious, feel free to check out the full changelog below.

We have a Matrix space if you would like to join in on discussions: https://matrix.to/#/#thunderapp:matrix.org

---

## Additions
* Updated create comment page UI and added language selector to comments by @hjiangsu in https://github.com/thunder-app/thunder/pull/1165
* Added ability to comment as a different user by @micahmo in https://github.com/thunder-app/thunder/pull/1352
* Added spoiler to markdown buttons by @micahmo and @hjiangsu in https://github.com/thunder-app/thunder/pull/1287, https://github.com/thunder-app/thunder/pull/1299
* Added support for selecting discussion languages by @iparks99 in https://github.com/thunder-app/thunder/pull/1289
* Added support for new sort types as defaults by @micahmo in https://github.com/thunder-app/thunder/pull/1298
* Added option to show user/community avatars by @iparks99 in https://github.com/thunder-app/thunder/pull/1318
* Added an option to hide thumbnails in feed by @iparks99 and @hjiangsu in https://github.com/thunder-app/thunder/pull/1302, https://github.com/thunder-app/thunder/pull/1326
* Added option to use profile picture for drawer by @micahmo in https://github.com/thunder-app/thunder/pull/1313, https://github.com/thunder-app/thunder/pull/1344
* Added support for customizable action colors by @micahmo in https://github.com/thunder-app/thunder/pull/1304
* Added selectable post/comment text by @micahmo in https://github.com/thunder-app/thunder/pull/1327
* Added notification debugging section by @micahmo in https://github.com/thunder-app/thunder/pull/1343
* Added confirmation message when unsubscribing from community by @micahmo in https://github.com/thunder-app/thunder/pull/1358
* Added experimental support for video player by @ggichure and @hjiangsu in https://github.com/thunder-app/thunder/pull/1081, https://github.com/thunder-app/thunder/pull/1369, and https://github.com/thunder-app/thunder/pull/1378
* Added sort options in instance explorer by @micahmo in https://github.com/thunder-app/thunder/pull/1317
* Added ability to perform empty search in Search by @micahmo in https://github.com/thunder-app/thunder/pull/1310

## Changes
* Show individual unread counts in inbox page by @micahmo in https://github.com/thunder-app/thunder/pull/1262
* Improved missing post snackbar message by @micahmo in https://github.com/thunder-app/thunder/pull/1291
* Improved spoiler appearance by @micahmo in https://github.com/thunder-app/thunder/pull/1286
* Adjusted markdown editor link dialog styling by @micahmo in https://github.com/thunder-app/thunder/pull/1307
* Adjusted padding to community sidebar by @micahmo in https://github.com/thunder-app/thunder/pull/1285
* Added loading page for posts/comments on slower connections by @micahmo in https://github.com/thunder-app/thunder/pull/1311
* Adjusted post/comment language selector styling by @hjiangsu in https://github.com/thunder-app/thunder/pull/1316
* Refactored post metadata and user/community chips by @hjiangsu in https://github.com/thunder-app/thunder/pull/1281
* Refactored post page quick actions by @hjiangsu in https://github.com/thunder-app/thunder/pull/1279
* Implemented core architecture to handle push notifications (APNs, UnifiedPush) by @hjiangsu in https://github.com/thunder-app/thunder/pull/1237
* Added retry action when initial feed loading fails by @micahmo in https://github.com/thunder-app/thunder/pull/1284
* Improved markdown appearance for block quotes, code blocks, and dividers by @micahmo in https://github.com/thunder-app/thunder/pull/1328
* Apply full name customizations to inbox mentions view by @micahmo in https://github.com/thunder-app/thunder/pull/1271
* Allow creating posts/comments with null language by @micahmo in https://github.com/thunder-app/thunder/pull/1338
* Optimistically mark replies as read by @micahmo in https://github.com/thunder-app/thunder/pull/1314
* Improved cross-post UI by @CTalvio in https://github.com/thunder-app/thunder/pull/1332
* Disabled Thunder's edge swipe to go back on Android by @hjiangsu in https://github.com/thunder-app/thunder/pull/1350
* Fixed various OLED color theme issues, community name format by @CTalvio in https://github.com/thunder-app/thunder/pull/1347
* Improved UX for notification settings by @micahmo in https://github.com/thunder-app/thunder/pull/1335
* Improved swipe gesture setting clarity with icons by @CTalvio in https://github.com/thunder-app/thunder/pull/1345
* Added loading page when opening notifications and increased size of notification icon by @micahmo in https://github.com/thunder-app/thunder/pull/1355
* Replaced score indicator icon by @CTalvio in https://github.com/thunder-app/thunder/pull/1341
* Improved image dimension fetching logic by @hjiangsu in https://github.com/thunder-app/thunder/pull/1369
* Improved the comment context jumping by @micahmo in https://github.com/thunder-app/thunder/pull/1319
* Improved account page and management UI by @CTalvio in https://github.com/thunder-app/thunder/pull/1354
* Improved selection of text when creating post/comment by @micahmo in https://github.com/thunder-app/thunder/pull/1359

## Fixes
* Fixed subscribe/favourite app bar action not updating UI by @micahmo in https://github.com/thunder-app/thunder/pull/1288
* Fixed comment load spinner showing incorrectly by @micahmo in https://github.com/thunder-app/thunder/pull/1278
* Fixed account page back button behaviour on Android by @hjiangsu in https://github.com/thunder-app/thunder/pull/1293
* Fixed issues with deleted comment content showing up by @micahmo in https://github.com/thunder-app/thunder/pull/1303
* Fixed search FAB primary single press action not triggering by @micahmo in https://github.com/thunder-app/thunder/pull/1308
* Fixed issue marking multiple replies as read by @micahmo in https://github.com/thunder-app/thunder/pull/1305
* Fixed issue with tablet mode images and full-height images by @hjiangsu in https://github.com/thunder-app/thunder/pull/1334
* Fixed pan/pinch triggering unintentional double tap zoom by @CTalvio in https://github.com/thunder-app/thunder/pull/1337
* Fixed issue popping notification page triggering twice by @micahmo in https://github.com/thunder-app/thunder/pull/1349
* Fixed viewing notifications as correct user by @micahmo in https://github.com/thunder-app/thunder/pull/1339
* Fixed FAB modal transparency and exit fade animation by @CTalvio in https://github.com/thunder-app/thunder/pull/1357

## Misc
* Integrate run on save scripts for VSCode by @micahmo in https://github.com/thunder-app/thunder/pull/1294
* Configure `i18n.ally` plugin by @ipcjs in https://github.com/thunder-app/thunder/pull/1277
* Apply linter suggested fix-ups by @micahmo in https://github.com/thunder-app/thunder/pull/1292
* Updated translations by @hjiangsu in https://github.com/thunder-app/thunder/pull/1362
* Updated instances by @github-actions in https://github.com/thunder-app/thunder/pull/1275
* Replaced scrollable_positioned_list with super_sliver_list by @hjiangsu in https://github.com/thunder-app/thunder/pull/1361
* Added test endpoint for UnifiedPush server and reduced UnifiedPush notification payload (in development) by @micahmo in https://github.com/thunder-app/thunder/pull/1364

## New Contributors
* @ipcjs made their first contribution in https://github.com/thunder-app/thunder/pull/1277
* @iparks99 made their first contribution in https://github.com/thunder-app/thunder/pull/1289

**Full Changelog**: https://github.com/thunder-app/thunder/compare/0.3.0...0.4.0

<hr /><em>This discussion was created from the release <a href='https://github.com/thunder-app/thunder/releases/tag/0.4.0'>v0.4.0</a>.</em>

**Comment by Spacellary** [comment]
To all contributors:
Thanks for the great work on Thunder!

**Reply by hjiangsu** [reply]
Thanks for the kind words!
